cluster-autoscaler/cloudprovider/openstack/openstack_cloud_provider.go

```diff
@@ -0,0 +1,177 @@
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+	"k8s.io/autoscaler/cluster-autoscaler/config"
+	"k8s.io/autoscaler/cluster-autoscaler/config/dynamic"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+	"k8s.io/klog"
+	"os"
+	"sync"
+)
+
+const (
+	// ProviderName is the cloud provider name for OpenStack
+	ProviderName = "openstack"
+)
+
+// OpenstackNodeGroup implements CloudProvider interface from cluster-autoscaler/cloudprovider module.
+type openstackCloudProvider struct {
+	openstackManager *OpenstackManager
+	resourceLimiter  *cloudprovider.ResourceLimiter
+	nodeGroups       []OpenstackNodeGroup
+}
+
+func buildOpenstackCloudProvider(openstackManager OpenstackManager, resourceLimiter *cloudprovider.ResourceLimiter) (cloudprovider.CloudProvider, error) {
+	os := &openstackCloudProvider{
+		openstackManager: &openstackManager,
+		resourceLimiter:  resourceLimiter,
+		nodeGroups:       []OpenstackNodeGroup{},
+	}
+	return os, nil
+}
+
+// Name returns the name of the cloud provider.
+func (os *openstackCloudProvider) Name() string {
+	return ProviderName
+}
+
+// NodeGroups returns all node groups managed by this cloud provider.
+func (os *openstackCloudProvider) NodeGroups() []cloudprovider.NodeGroup {
+	groups := make([]cloudprovider.NodeGroup, len(os.nodeGroups))
+	for i, group := range os.nodeGroups {
+		groups[i] = &group
+	}
+	return groups
+}
+
+// AddNodeGroup appends a node group to the list of node groups managed by this cloud provider.
+func (os *openstackCloudProvider) AddNodeGroup(group OpenstackNodeGroup) {
+	os.nodeGroups = append(os.nodeGroups, group)
+}
+
+// NodeGroupForNode returns the node group that a given node belongs to.
+//
+// Since only a single node group is currently supported, the first node group is always returned.
+func (os *openstackCloudProvider) NodeGroupForNode(node *apiv1.Node) (cloudprovider.NodeGroup, error) {
+	// TODO: wait for magnum nodegroup support
+	return &(os.nodeGroups[0]), nil
+}
+
+// Pricing is not implemented.
+func (os *openstackCloudProvider) Pricing() (cloudprovider.PricingModel, errors.AutoscalerError) {
+	return nil, cloudprovider.ErrNotImplemented
+}
+
+// GetAvailableMachineTypes is not implemented.
+func (os *openstackCloudProvider) GetAvailableMachineTypes() ([]string, error) {
+	return []string{}, nil
+}
+
+// NewNodeGroup is not implemented.
+func (os *openstackCloudProvider) NewNodeGroup(machineType string, labels map[string]string, systemLabels map[string]string,
+	taints []apiv1.Taint, extraResources map[string]resource.Quantity) (cloudprovider.NodeGroup, error) {
+	return nil, cloudprovider.ErrNotImplemented
+}
+
+// GetResourceLimiter returns resource constraints for the cloud provider
+func (os *openstackCloudProvider) GetResourceLimiter() (*cloudprovider.ResourceLimiter, error) {
+	return os.resourceLimiter, nil
+}
+
+// Refresh is called before every autoscaler main loop.
+//
+// Currently only prints debug information.
```

Comment:
Is there a plan to have this Refresh function gather all API resources to avoid GET calls during the scaling operations? (similar to other cloudproviders)

Reply:
@scott-chaney if scaling would only need to happen fairly infrequently, I think that this would just be unnecessarily making a call to magnum and two calls to heat every 10 seconds. Is there any other benefit to having the values stored other than not having to get them during the scaling?

Reply:
@tghartland It might be less apparent since the NodeGroups and NodeGroupForNode functions aren't hitting the API yet (they will definitely need to when heterogeneous clusters are supported, and they probably should be checking now to help alleviate some error scenarios) but even when there is no scale event needed, I see several dozen interface method calls over each autoscaler loop (after the refresh).

Caching on first lookup and purging on refresh could work but it's messier (just IMO) and something like NodeGroupForNode which could be satisfied in one API calls would now take as many as there are nodes in the cluster.

Once NodeGroups and NodeGroupForNode are implemented it might be worth revisiting.

Reply:
@scott-chaney I was working yesterday on part of DeleteNodes where I noticed that the first check comparing the expected new size and the minimum size was using a target size that was potentially out of date (if the user had manually resized their cluster, for instance).

What I ended up with is something like what you described, where the first DeleteNodes call gets and stores the value and then the other simultaneous calls are able to use the value.
https://github.com/kubernetes/autoscaler/pull/1690/commits/08787e951ed662b1be1aa19db1dea8b27473133d#diff-68f03f546ccae8f6141466056b52fabdR139

The major problem I was dealing with in that commit was making sure that only one of the simultaneous deleteNodes calls was the one to do the caching, so this definitely would be made easier by doing the caching elsewhere, like in Refresh.

The heat manager only has two functions (nodeGroupSize and getClusterStatus) that call clusters.Get, so it would be simple to separate that into its own getCluster method that both call. Then that function can use whatever kind of caching is available.

Reply:
@scott-chaney I wrote some code for caching the getCluster result, but I think it's because I was aiming at that one specific resource that it felt a bit hacky, instead of properly caching the full state of the cluster and stacks.

I tried both with caching on Refresh and purging on Refresh then caching when needed, but both still felt unnecessary at the moment. Most of the cluster/stack API calls in the scale up and scale down are checking for the status changes anyway, so they were having to purposefully clear the cache to force a fresh result.

NodeGroupForNode does get called a lot, but for that I think it would be convenient to cache a mapping of node -> nodegroup to check first, and then rebuild that map after a scale up/down, or if a node is not found in the map (with a timeout to prevent this happening many times at once). I'll revisit this and general caching once there are actual node groups to work with.